contracts/pol/CarbonPOL.sol

```diff
@@ -0,0 +1,303 @@
+     * @dev performs contract-specific initialization
+     */
+    function __CarbonPOL_init_unchained() internal onlyInitializing {
+        // set market price multiplier to 2x
+        _setMarketPriceMultiply(2);
+        // set price decay half-life to 10 days
+        _setPriceDecayHalfLife(10 days);
+    }
+
+    /**
+     * @dev authorize the contract to receive the native token
+     */
+    receive() external payable {}
+
+    /**
+     * @dev validate token
+     */
+    modifier validToken(Token token) {
+        _validToken(token);
+        _;
+    }
+
+    /**
+     * @dev validate price
+     */
+    modifier validPrice(Price memory price) {
+        _validPrice(price);
+        _;
+    }
+
+    /**
+     * @inheritdoc Upgradeable
+     */
+    function version() public pure override(IVersioned, Upgradeable) returns (uint16) {
+        return 1;
+    }
+
+    /**
+     * @notice sets the market price multiply
+     *
+     * requirements:
+     *
+     * - the caller must be the admin of the contract
+     */
+    function setMarketPriceMultiply(
+        uint32 newMarketPriceMultiply
+    ) external onlyAdmin greaterThanZero(newMarketPriceMultiply) {
+        _setMarketPriceMultiply(newMarketPriceMultiply);
+    }
+
+    /**
+     * @notice sets the price decay half-life
+     *
+     * requirements:
+     *
+     * - the caller must be the admin of the contract
+     */
+    function setPriceDecayHalfLife(
+        uint32 newPriceDecayHalfLife
+    ) external onlyAdmin greaterThanZero(newPriceDecayHalfLife) {
+        _setPriceDecayHalfLife(newPriceDecayHalfLife);
+    }
+
+    /**
+     * @notice enable trading for a token and set the initial price
+     *
+     * requirements:
+     *
+     * - the caller must be the admin of the contract
+     */
+    function enableTrading(Token token, Price memory price) external onlyAdmin validPrice(price) {
+        _tradingStartTimes[token] = uint32(block.timestamp);
+        _initialPrice[token] = price;
+        emit TradingEnabled(token, price);
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function marketPriceMultiply() external view returns (uint32) {
+        return _marketPriceMultiply;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function priceDecayHalfLife() external view returns (uint32) {
+        return _priceDecayHalfLife;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function tradingEnabled(Token token) external view returns (bool) {
+        return _tradingEnabled(token);
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function expectedTradeReturn(Token token, uint128 ethAmount) external view returns (uint128 tokenAmount) {
+        // return 0 if trading is not enabled for token
+        if (!_tradingEnabled(token)) {
+            return 0;
+        }
+        Price memory currentPrice = tokenPrice(token);
+        // revert if ethAmount in price is equal to 0
+        if (currentPrice.ethAmount == 0) {
```

Comment:
To be on the safe side and future proof, I'd suggest using `_validPrice` instead.

Reply:
Implemented suggestion